p2p/src/sync/peer_v2.rs

```diff
@@ -708,14 +707,28 @@ where
         // The code below will set it again if needed.
         self.peer_activity.set_expecting_blocks_since(None);
 
-        // TODO: here we allow the peer to send blocks out of order. This means that we may receive
-        // an orphan block first and then its parent. Because of this, the NewTipReceived event
-        // generated below may not contain the actual tip. It doesn't seem like it may lead
-        // to problems at this moment, but it's still better to be more strict about it.
-        if self.incoming.requested_blocks.take(&block.get_id()).is_none() {
-            return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
-                "block response".to_owned(),
-            )));
+        if self.incoming.requested_blocks.front().is_some_and(|id| id == &block.get_id()) {
+            self.incoming.requested_blocks.pop_front();
+        } else {
+            let idx = self.incoming.requested_blocks.iter().position(|id| id == &block.get_id());
+            // Note: we treat wrongly ordered blocks in the same way as unsolicited ones, i.e.
+            // we don't remove their ids from the list.
+            if idx.is_some() {
+                return Err(P2pError::ProtocolError(
+                    ProtocolError::BlocksReceivedInWrongOrder {
+                        expected_block_id: *self
+                            .incoming
+                            .requested_blocks
+                            .front()
+                            .expect("The deque is known to be non-empty"),
+                        actual_block_id: block.get_id(),
+                    },
+                ));
+            } else {
+                return Err(P2pError::ProtocolError(
+                    ProtocolError::UnsolicitedBlockReceived(block.get_id()),
+                ));
+            }
```

Comment:
I'm not sure whether we need this distinction between `BlocksReceivedInWrongOrder` and `UnsolicitedBlockReceived`, provided that it's "informational" only (initially I was planning to treat them differently, that's why I introduced this separate `BlocksReceivedInWrongOrder` error)

Reply:
I think it's fine to keep the distinction. It may help to detect issues if we miss something.